plugin/src/main/java/com/denizenscript/denizen/events/entity/ProjectileHitScriptEvent.java

```diff
@@ -0,0 +1,157 @@
+package com.denizenscript.denizen.events.entity;
+
+import com.denizenscript.denizen.events.BukkitScriptEvent;
+import com.denizenscript.denizen.objects.EntityTag;
+import com.denizenscript.denizen.objects.LocationTag;
+import com.denizenscript.denizen.objects.MaterialTag;
+import com.denizenscript.denizen.utilities.BukkitImplDeprecations;
+import com.denizenscript.denizen.utilities.implementation.BukkitScriptEntryData;
+import com.denizenscript.denizencore.objects.ObjectTag;
+import com.denizenscript.denizencore.scripts.ScriptEntryData;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.entity.ProjectileHitEvent;
+
+public class ProjectileHitScriptEvent extends BukkitScriptEvent implements Listener {
+
+    // <--[event]
+    // @Events
+    // <projectile> hits
+    //
+    // @Synonyms entity shoots
+    //
+    // @Switch entity:<entity> to only process the event if an entity got hit, and it matches the specified EntityTag matcher.
+    // @Switch block:<block> to only process the event if a block got hit, and it matches the specified LocationTag matcher.
+    // @Switch shooter:<entity> to only process the event if the projectile was shot by an entity, and it matches the specified EntityTag matcher.
+    //
+    // @Group Entity
+    //
+    // @Location true
+    //
+    // @Cancellable true
+    //
+    // @Triggers When a projectile hits a block or an entity.
+    //
+    // @Context
+    // <context.projectile> returns an EntityTag of the projectile.
+    // <context.hit_entity> returns an EntityTag of the entity that was hit, if any.
+    // <context.hit_block> returns a LocationTag of the block that was hit, if any.
+    // <context.hit_face> returns a LocationTag vector of the hit normal (like '0,1,0' if the projectile hit the top of a block).
+    // <context.shooter> returns an EntityTag of the entity that shot the projectile, if any.
+    //
+    // @Player when the entity that was hit is a player, or when the shooter is a player if no entity was hit.
+    //
+    // @NPC when the entity that was hit is a npc, or when the shooter is a npc if no entity was hit.
+    //
+    // -->
+
+    public ProjectileHitScriptEvent() {
+        registerCouldMatcher("<projectile> hits (<material>)");
+        registerCouldMatcher("<projectile> hits (<entity>)");

```

Comment:
One of these two need to at least be non-optional, as otherwise `<projectile> hits` is registered twice